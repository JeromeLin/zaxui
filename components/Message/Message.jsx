import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Icon from '../Icon';

class Message extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      visible: props.visible || true,
    };
  }

  render() {
    const { prefixCls, className, theme, size, icon, hasArrow, hasClosable, onClick, children } = this.props;

    const classes = classnames({
      [`${prefixCls}`]: true,
      [className]: !!className,
      [`theme-${theme}`]: !!theme,
      [`size-${size}`]: !!size,
    });

    const iconRender = icon
      ? <div className={`${prefixCls}-icon`}>{icon}</div>
      : null;

    const renderCloseIcon = hasClosable
      ? <Icon type="wrong" onClick={() => { this.setState({ visible: false }); }} />
      : null;

    const renderArrow = hasArrow
      ? <Icon type="arrow-right" />
      : null;

    const noFooter = !hasClosable && !hasArrow;

    return this.state.visible ? (
      <div className={classes} onClick={renderArrow && onClick}>
        <div className={`${prefixCls}-header`}>{iconRender}</div>
        <div className={`${prefixCls}-body`}>{children}</div>
        {
          !noFooter
            ? <div className={`${prefixCls}-footer`}>{renderArrow}{renderCloseIcon}</div>
            : null
        }
      </div>
    ) : null;
  }
}

Message.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  theme: PropTypes.oneOf(['default', 'primary', 'info', 'success', 'warning', 'error']),
  size: PropTypes.oneOf(['lg']),
  visible: PropTypes.bool,
  icon: PropTypes.element,
  hasArrow: PropTypes.bool,
  hasClosable: PropTypes.bool,
  onClick: PropTypes.func,
};

Message.defaultProps = {
  prefixCls: 'za-message',
  className: null,
  theme: 'primary',
  size: null,
  visible: true,
  icon: null,
  hasArrow: false,
  hasClosable: false,
  onClick() {},
};

export default Message;
