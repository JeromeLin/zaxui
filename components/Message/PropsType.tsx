export default interface PropsType {
  theme?: 'default' | 'primary' | 'info' | 'success' | 'warning' | 'error';
  size?: 'lg';
  icon?: any;
  hasArrow?: boolean;
  hasClosable?: boolean;
  onClick?: () => void;
}
