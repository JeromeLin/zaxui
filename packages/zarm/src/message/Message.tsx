import { ArrowRight as ArrowRightIcon, Close as CloseIcon } from '@zarm-design/icons';
import classnames from 'classnames';
import type { HTMLAttributes, MouseEvent } from 'react';
import React, { forwardRef, useState } from 'react';
import type { BaseMessageProps } from './interface';

export interface MessageProps extends BaseMessageProps, HTMLAttributes<HTMLDivElement> {
  prefixCls?: string;
  onClick?: (e: MouseEvent<HTMLDivElement>) => void;
}

const Message = forwardRef<HTMLDivElement, MessageProps>((props, ref) => {
  const {
    prefixCls = 'za-message',
    className,
    theme = 'primary',
    size,
    icon,
    hasArrow = false,
    closable = false,
    children,
    onClick,
    ...restProps
  } = props;
  const [visible, setVisible] = useState(true);

  const classes = classnames(prefixCls, className, {
    [`${prefixCls}--${theme}`]: !!theme,
    [`${prefixCls}--${size}`]: !!size,
    [`${prefixCls}--link`]: !!hasArrow,
  });

  function handleClick(e: MouseEvent<HTMLDivElement>) {
    if (!hasArrow) {
      e.preventDefault();
      return;
    }
    onClick?.(e);
  }

  function handleClose() {
    setVisible(false);
  }

  const iconRender = icon && <div className={`${prefixCls}__icon`}>{icon}</div>;
  const closeIconRender = closable && <CloseIcon onClick={handleClose} />;
  const arrowRender = hasArrow && <ArrowRightIcon />;
  const noFooter = !closable && !hasArrow;

  if (visible) {
    return (
      <div ref={ref} className={classes} onClick={handleClick} {...restProps}>
        <div className={`${prefixCls}__header`}>{iconRender}</div>
        <div className={`${prefixCls}__body`}>{children}</div>
        {!noFooter && (
          <div className={`${prefixCls}__footer`}>
            {arrowRender}
            {closeIconRender}
          </div>
        )}
      </div>
    );
  }

  return null;
});

export default Message;
