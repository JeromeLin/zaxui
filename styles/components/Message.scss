$prefixCls: za-message;

@mixin message-theme($color, $background, $border) {
  color: $color;
  background-color: $background;
  border-color: $border;
  @include onepx($border-color: $border);
}

.#{$prefixCls} {
  display: flex;
  align-items: center;
  position: relative;
  padding: r(5) r(8);
  min-height: r(30);
  font-size: r(12);
  @include message-theme($button-default-border, $button-default-activebg, $button-default-border);

  &.theme-primary {
    @include message-theme($button-primary-border, $button-primary-activebg, $button-primary-border);
  }

  &.theme-info {
    @include message-theme($button-info-border, $button-info-activebg, $button-info-border);
  }

  &.theme-success {
    @include message-theme($button-success-border, $button-success-activebg, $button-success-border);
  }

  &.theme-warning {
    @include message-theme($button-warning-border, $button-warning-activebg, $button-warning-border);
  }

  &.theme-error {
    @include message-theme($button-error-border, $button-error-activebg, $button-error-border);
  }
}
