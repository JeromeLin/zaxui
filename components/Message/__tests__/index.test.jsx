import React from 'react';
import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import Message from '../index';

describe('Message', () => {
  it('renders correctly', () => {
    const wrapper = render(<Message>foo</Message>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
