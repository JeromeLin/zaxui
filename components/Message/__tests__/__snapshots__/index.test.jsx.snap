// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message hasArrow 1`] = `
<div
  className="za-message theme-primary"
  onClick={[Function]}
>
  <div
    className="za-message-header"
  />
  <div
    className="za-message-body"
  >
    foo
  </div>
  <div
    className="za-message-footer"
  >
    <Icon
      className={null}
      prefixCls="za-icon"
      theme={null}
      type="arrow-right"
    />
  </div>
</div>
`;

exports[`Message hasClosable 1`] = `
<div
  className="za-message theme-primary"
  onClick={null}
>
  <div
    className="za-message-header"
  />
  <div
    className="za-message-body"
  >
    foo
  </div>
  <div
    className="za-message-footer"
  >
    <Icon
      className={null}
      onClick={[Function]}
      prefixCls="za-icon"
      theme={null}
      type="wrong"
    />
  </div>
</div>
`;

exports[`Message icon 1`] = `
<div
  class="za-message theme-primary"
>
  <div
    class="za-message-header"
  >
    <div
      class="za-message-icon"
    >
      <img
        alt=""
        src="https://zhongantecheng.github.io/zarm/images/state.18e78939.png"
      />
    </div>
  </div>
  <div
    class="za-message-body"
  >
    foo
  </div>
</div>
`;

exports[`Message renders correctly 1`] = `
<div
  class="za-message theme-primary"
>
  <div
    class="za-message-header"
  />
  <div
    class="za-message-body"
  >
    foo
  </div>
</div>
`;

exports[`Message theme 1`] = `
<div
  class="za-message theme-error"
>
  <div
    class="za-message-header"
  />
  <div
    class="za-message-body"
  >
    foo
  </div>
</div>
`;
